tests/test_mutations.py

```diff
@@ -472,47 +705,288 @@ def test_keep_mutation_parent_zero_rate(self):
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.rate[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_positions = [s.position for s in self.sites]
+                    index = np.searchsorted(site_positions, position)
+                    if (index == len(self.sites)
+                       or site_positions[index] != position):
+                        # new site
+                        ancestral_state = self.alleles[self.pick_allele(rng)]
+                        site = SiteItem(
+                                position=position,
+                                ancestral_state=ancestral_state,
+                                metadata=b'',
+                                mutations=[])
+                        self.sites.insert(index, site)
+                    else:
+                        site = self.sites[index]
+                    # keep mutations sorted in decreasing order by time
+                    mut_times = [m.time for m in site.mutations]
+                    mut_index = np.searchsorted(-1 * mut_times, -1 * time)
+                    mutation = MutationItem(
+                            site=tskit.NULL,
+                            node=edge.child,
+                            derived_state="Y",
+                            parent=tskit.NULL,
+                            metadata=b'',
+                            time=time,
+                            new=True,
+                            id=tskit.NULL)
+                    site.mutations.insert(mut_index, mutation)
+                    # Sample from this to duplicate the behaviour in the C implementation
+                    rng.uniform_int(1)
+                index += 1
+                left = right
+
+    def pick_allele(self, rng):
+        u = rng.flat(0, 1)
+        j = 0
+        while u > self.alphabet_distribution[j]:
+            u -= self.alphabet_distribution[j]
+            j += 1
+        return j
+
+    def transition_allele(self, j, rng):
+        u = rng.flat(0, 1)
+        k = 0
+        while u > self.mutation_matrix[j][k]:
+            u -= self.mutation_matrix[j][k]
+            k += 1
+        return k
+
+    def choose_alleles(self, tables, rng):
```

Comment:
I had something different in mind - we use compute mutation parents and we just build the **mutation** tree at each site. I.e.
```python
root = MutationTreeNode()
nodes = {-1: root}
for mutation in site_mutations:
     node =  MutationTreeNode(mutation)
     nodes[mutation.id] = MutationTreeNode(mutation)
     node.parent = nodes[mutation.parent]
     nodes[mutation.parent].children.append(node)
```

Then, traversing the tree with states is trivial:
```python
state = # if it's new, choose an allele, otherwise use original state
stack = [(root, state)]
while len(stack) > 0:
    mtn, state = stack.pop()
    for child_mtn in mtn.children():
          # Choose a new state for the child according to distribution, etc
          new_state = something
          stack.push((child_mtn, new_state))
```

So, we use the compute_mutation_parents method that we've tuned for perf already, and the task becomes one of doing a top-down traversal of the **mutation tree**, which should contain a handful of nodes, most of the time. We should be able to do this in Python, even (but, maybe not). Easy enough to do in C anyway.

Reply:
Right - well, the reason I stopped using `compute_mutation_parents` is that it requires the tables, and so I was having to build the tables, compute the parents, and then go back and remove some rows of the tables once I had chosen the alleles, which was feeling ugly and inefficient. I could modify the code from `compute_mutation_parents` to build the tree first, then do like this - but, that's basically what I did to write this function. So - I'm not sure about this?

Reply:
It's a bit unfortunate that we have to build the tables and then go back and update the site and mutation tables again I agree, but it's not terrible right? All we're doing really is updating the ancestral and derived state columns and that's easy/efficient.

I don't understand why we'd need to remove rows - is that because you get rid of mutations that happen to give the same state when we condition on some already-existing mutation? If so, we can always just build the mutation table row-by-row afterwards, it's not terrible. I'd make this as clean and simple as we can first, and then worry about making it more efficient afterwards.

Reply:
Another possibility would be to rejection sample any mutation configurations that don't make sense, I guess?

Reply:
I'm going to push back a bit here. Let's see:

> Another possibility would be to rejection sample any mutation configurations that don't make sense, I guess?

We really *do* want to possibly be sampling mutations that don't do anything. It makes everything much cleaner, as it allows us to separate out throwing down mutations from picking states; otherwise, mutation rate depends on previous state, and there is lots more bookkeeping. 

> why we'd need to remove rows - is that because you get rid of mutations that happen to give the same state when we condition on some already-existing mutation?

Right, this is why we need to remove mutations (if they don't change the state, because tskit requires that derived state differs from the previous state), and thus maybe sites (if all of a site's mutations are removed).

> It's a bit unfortunate that we have to build the tables and then go back and update the site and mutation tables again I agree, but it's not terrible right?

Well, that's what I thought at first as well, but then I coded it up and it *was* terrible, and after a good bit of thought I decided that this way was much nicer. This was totally not obvious until my head was really in it. I'll put some more work into simplifying the python algorithm, though - maybe there's some more refactoring that can be done.

Reply:
I see, keeping the 'silent' mutations makes the maths easier, so fair enough. Well I'd still advocate for doing the state changes on the mutation tree - it's just so pretty! Doing the second table update row-by-row shouldn't be *too* bad (plus, that's what we'll be doing in C anyway - no advantage to mucking around with set_columns there.)

Reply:
> Doing the second table update row-by-row shouldn't be too bad

It is do-able, sure! But, the current approach is also not too bad - and, I'm pretty sure, better than updating the tables twice.

Reply:
OK, I'll shut up then :smile: Ping me when you're happy with it.

Reply:
You may well be right, I just don't think so right now. =)

Reply:
I've moved 'choosing alleles' into a separate function called once each tree is built (see next commit). And, then I spent quite a while refactoring "build the tree of mutations" out from "choose the alleles", like this:
```
                site = self.sites[positions[j]]
                # sort mutations by (decreasing time, increasing insertion order)
                site.mutations.sort(key = lambda m: (-1 * m.time, m.parent))
                mut_parent = [tskit.NULL for _ in site.mutations]
                # Build the tree of mutations:
                # bottom_mutation maps from nodes to the most recent
                #  mutation on the edge above the node, if any
                bottom_mutation = {}
                for k in range(len(site.mutations)):
                    mut = site.mutations[k]
                    u = mut.node
                    while u != tskit.NULL and u not in bottom_mutation:
                        u = tree_parent[u]
                    if u != tskit.NULL:
                        assert u in bottom_mutation
                        mut_parent[k] = bottom_mutation[u]
                    bottom_mutation[mut.node] = k

                num_mutations = self.choose_alleles(mut_parent, site, mutation_id, rng)
```
... which is very nice, **except** that inside `choose_alleles` we have to traverse up the tree of mutations to avoid mutations that have to be dropped. So, we end up having tree traversals in both parts of the code (one traversing up the tree of nodes to produce the tree of mutations, but then traversing up the tree of mutations), instead of just a single tree traversal. So, I've backed out those changes but am happy to pass them along if you are curious.